### ISSUE TITLE
Issues in referencing multiple items pointing to the same IRI

### ISSUE THREAD

**Issue by mprorock** [w3c-ccg/traceability-vocab#88]
@Therecanbeonlyone1969 
There is a trick to this that i used in [IntentToSell](https://github.com/w3c-ccg/traceability-vocab/blob/main/docs/schemas/IntentToSell.json)

effectively in the comment section where you are defining the `@ id` you have to have a unique value as it uses a struct to map over when building the json-ld from the schema, the easy way to do this is to add a query param that is meaningless, but unique within the object (similar to cache busting)

edit:
```
"seller": {
            "$comment": "{\"term\": \"seller\", \"@id\": \"https://w3id.org/traceability?1#Entity\"}",
            "title": "Seller",
            "description": "Entity that is declaring their intention for a product to be used at a different location than where it was created.",
            "$ref": "https://w3id.org/traceability/schemas/Entity.json"
        },
        "purchaser": {
            "$comment": "{\"term\": \"purchaser\", \"@id\": \"https://w3id.org/traceability?2#Entity\"}",
            "title": "Purchaser",
            "description": "Entity that is expected to receive some amount of the product, though possibly a reseller.",
            "$ref": "https://w3id.org/traceability/schemas/Entity.json"
        },
```
note the `?1` and `?2` in seller and purchaser, even though both resolve to `Entity`

I have validated this on the json-ld org's playground as well.  Big thing to be aware of is making sure you are using a query string that isgoing to be ignored by the hosting page, and in the case of stuff resolving to w3id.org (or other gh pages stuff like it), make sure the query string comes before the hash

_Originally posted by @mprorock in https://github.com/w3c-ccg/traceability-vocab/pull/79#discussion_r573934958_

**Comment by mprorock** [comment]
@OR13 ultimately i think it will be better to allow have the json schema take an `$jld-id`, `$jld-term`, etc but we will have to approach that carefully to make sure that you can use the schema on it's own directly without forcing use of the json-ld.  `optional` comes to mind, but would be super verbose, but compared to having to escape a bunch of stuff in a comment string, would likely be more clear.

for example, something like this would potentially work and we could standardize:

```
"seller": {
    "title": "Seller",
    "description": "Entity that is declaring their intention for a product to be used at a different location than where it was created.",
    "$ref": "https://w3id.org/traceability/schemas/Entity.json"
    "json-ld-context" : {
        "type": "object",
        "optional": true,
        "properties": {
                "@id": "https://w3id.org/traceability#Entity",
                "term": "seller"
        }
    }
},
```

I think the right work area for this is in the [linked data processor](https://github.com/transmute-industries/linked-data/blob/main/packages/jsonld-schema/src/schemasToContext.ts) you authored, and then to use that library here if we can align that library with a standard that will work with this and other schema -> ld -> schema type use cases

Thoughts?

**Comment by OR13** [comment]
The bug is that the JSON-LD generator does not handle term reuse...  this is a problem since the whole point of this is to be able to generate JSON-LD from JSON Schema... and the last thing we want to do is discourage term reuse.

The solution is to update the logic so that JSON Schema types and and JSON-LD types can include overlapping  / duplicate term IRIs.

If you want to test this in isolation feel free to use https://github.com/transmute-industries/linked-data/tree/main/packages/jsonld-schema

and write a small unit test that reproduces the bug before attempting a fix.

**Comment by OR13** [comment]
DO NOT DO THIS

```
https://w3id.org/traceability?2#Entity\
```

You are making IRIs that are a really bad idea to be signed over...

**Comment by mprorock** [comment]
agree - this however should not block us from having stuff up and going.  We can always add a post process that strips query strings from the json-ld that match a certain documented pattern to avoid signing issues

**Comment by Therecanbeonlyone1969** [comment]
@mprorock agree with @OR13 ... we need to be able to reuse terms and do not have them overwritten. The generator should be readily fixable. There must be an `@id` check in the code somewhere. We need to alter that bit of code. Question is who fixes it, because this is an issue affecting everyone. There is another issue ... i will open in a 2nd that is also high priority that needs to be fixed asap as well. That one is in the JSON-LD checker. I am happy to work on both if I do not have to do it alone.

**Comment by mprorock** [comment]
@Therecanbeonlyone1969 fully agree - as to who alters the code, we need to hop on it quickly.  you are welcome to do so if you have bandwidth and a good solution

**Comment by OR13** [comment]
any update on this? AFAIK its a critical issue, would love to see a solution...

**Comment by Therecanbeonlyone1969** [comment]
@OR13 ... since I need to decompose a lot of large complex schemas into many smaller schemas, because the context generator only resolves and thus adds top-level properties to the context, I am in schema decomposition hell at the moment and have no cycles to look at the code at the moment ... @mprorock , does anyone on your team have cycles?

**Comment by OR13** [comment]
@Therecanbeonlyone1969 I suggest small PRs that change on thing at a time... and that are easy to merge with confidence.

**Comment by Therecanbeonlyone1969** [comment]
> @Therecanbeonlyone1969 I suggest small PRs that change on thing at a time... and that are easy to merge with confidence.

next PR will be 1 VC and 1 evidence doc with tests

**Comment by OR13** [comment]
Fix is up here: https://github.com/transmute-industries/linked-data/pull/15

**Comment by OR13** [comment]
Now we need to remove all the hacking ?v=N to from URLs in order to close this issue.

**Comment by Therecanbeonlyone1969** [comment]
> Now we need to remove all the hacking ?v=N from URLs in order to close this issue.

@OR13 ... I never used those -- only resolvable links, but I abused some references a little bit where just using e.g. schema.org/identifier would have been the most accurate IRI ... I will make the e-commerce ones more accurate next week

**Comment by mprorock** [comment]
Guilty.  Will resolve

Michael Prorock
CTO, Founder
mesur.io

**Comment by OR13** [comment]
@mprorock any update on this?

**Comment by mprorock** [comment]
Just ran some tests this morning and am still hitting some issues when duplicate ids are used in the comment tag.  Digging in a bit more to see why

**Comment by OR13** [comment]
@mprorock I have a unit test for this issue here: https://github.com/transmute-industries/verifiable-data/blob/main/packages/jsonld-schema/src/__tests__/duplicate-terms.test.ts

but feel free to open a PR with a counter example if my fix didn't cover it all.

**Comment by mprorock** [comment]
Cool, like I said I am stepping the code through to make sure it is not on my end

**Comment by Therecanbeonlyone1969** [comment]
@mprorock I had no issues anymore. cc @OR13

**Comment by OR13** [comment]
On the call: In order to close this out, we need to remove all query string hacks from all IRIs in the JSON-LD generated context.

**Comment by mkhraisha** [comment]
I think i still need to clean those up on our end.

**Comment by TallTed** [comment]
_**tl;dr:** There are many problems brought to light by this issue.  Please take the time to read this very long comment, as the sooner these errors are addressed, the sooner good data and tests can be deployed._

-----

I do not think the title of this issue reflects its current evolved state (and I'm not sure it ever accurately reflected the issue being discussed, including in the original post).

Multiple items should be permitted to have the same type a/k/a class, in which case, their type should be designated by the same URI.  I read the title of this issue as saying that there are problems with the preceding sentence.  Reading the rest of the issue, I see that's not exactly the issue, but there appear to be many misconceptions/misunderstandings driving both code and tests.

I see this very problematic chunk of JSON --

```
"seller": {
            "$comment": "{\"term\": \"seller\", \"@id\": \"https://w3id.org/traceability?1#Entity\"}",
            "title": "Seller",
            "description": "Entity that is declaring their intention for a product to be used at a different location than where it was created.",
            "$ref": "https://w3id.org/traceability/schemas/Entity.json"
        },
"purchaser": {
            "$comment": "{\"term\": \"purchaser\", \"@id\": \"https://w3id.org/traceability?2#Entity\"}",
            "title": "Purchaser",
            "description": "Entity that is expected to receive some amount of the product, though possibly a reseller.",
            "$ref": "https://w3id.org/traceability/schemas/Entity.json"
        },
```

0. `@id` appears to be being misused, apparently in place of `rdf:type`.  Looking at [duplicate-terms.test.ts](https://github.com/transmute-industries/verifiable-data/blob/main/packages/jsonld-schema/src/__tests__/duplicate-terms.test.ts), I find this is absolutely the case --
   ```
          firstName: {
            '@id': 'https://schema.org/name',
          },
          lastName: {
            '@id': 'https://schema.org/name',
   ```
   * The *@id* of each object should be a unique IRI.

   * The *@value* is where you put the obvious string (or whatever datatype).

   * The *@type* gets the class of the value, i.e., `https://schema.org/name` here -- though I do not recommend settig both of *these* values to that same rdf:type, because there's no good way to get back to which value comes before the other.

   * Also, PLEASE learn from this article, [_Falsehoods Programmers Believe About Names – With Examples_](https://shinesolutions.com/2018/01/08/falsehoods-programmers-believe-about-names-with-examples/)

1. Structured data that is meant to be interpreted as such should not be wedged into an unstructured field of a structured data form!  The `Comment` should not be abused in this way.

2. A "Seller" might well not declare any "intention for a product to be used at a different location than where it was created."  There are endless scenarios where a sold item might be consumed at the same location as where it was created (or purchased).  There are many where the "seller" might be one department and the "buyer" another department, both within the same business entity, and at the same street address -- though perhaps their "physical location" might include the floor of a building, or some other sub-division location.  There are many sales, sometimes several in a row, where the item being sold never moves from a warehouse location or the like.

   The "Seller" is declaring their intention to transfer ownership of some item(s) to the "Buyer" (used more commonly than "Purchaser"), in exchange for something else (currency, other input materials, other input or output goods, etc.).  That should be all that's discussed in the description of this transactional actor/entity.

3. Why is the "Buyer" called out as "though possibly a reseller"?  Certainly a Buyer may or may not actually take physical possession of the goods (or anything else!  e.g., land, mining rights, airspace rights, easement/right-of-way) they're purchasing, whether or not they're purchasing with the intent to resell, or find they've overbought some supply item, etc.

4. What is the intent of this URI structure, `https://w3id.org/traceability?2#Entity\`?  Among other things
   * the trailing backslash is poison
   * vocab URIs that include query segments are likely to put excessive demands on both the redirector and the eventual server

**Comment by OR13** [comment]
The query strings need to be removed.

**Comment by rhofvendahl** [comment]
I'd be happy to tackle this.

**Comment by rhofvendahl** [comment]
I ran into an issue removing the query strings.

Many of the query strings can be removed without problem, but this is only because some classes don't actually have properties with duplicate IDs. If I remove the query strings for IntentToSell which does have properties with the same ID ("seller" and "purchaser") or change two IDs on any two same-class properties to be the same, I get an error that includes:
```
traceability-schemas: /Users/russell/traceability-vocab/packages/traceability-schemas/scripts/getContextFromIntermediate.js:17
traceability-schemas:         [classProperty.$comment.term]: {
traceability-schemas:                                 ^
traceability-schemas: TypeError: Cannot read property 'term' of undefined
```

My understanding is that @OR13's changes made it so that when properties with duplicate IDs are added to the classDefinition.classProperties object, they are added as an array of properties rather than a single property (i.e. classDefinition.classProperties[ID link] returns an Array. Commit with changes [here](https://github.com/w3c-ccg/traceability-vocab/pull/146/commits/1cef8985d00652a8f2d6cdd32f139d5ed0fbcda4), code commit references [here](https://github.com/transmute-industries/verifiable-data/blob/c80cab9abe4db478add16b14837ba9a3afc3a70f/packages/jsonld-schema/src/schemasToContext.ts)).

It looks as though the getContextFromIntermediate function (in getContextFromIntermediate.js) tries to access .$comment.term directly off of classDefinition.classProperty[ID link], expecting an object rather than an array. It should be easy to check for an array first here and process those arrays accordingly, and I'd be happy to make this change. However, I don't know if there are other areas farther along in the processing pipeline that might be affected.

**Comment by OR13** [comment]
@rhofvendahl its possible the the current code in this repo, does not implement the fix... does this text and library address your concern? https://github.com/transmute-industries/verifiable-data/blob/main/packages/jsonld-schema/src/__tests__/duplicate-terms.test.ts

**Comment by rhofvendahl** [comment]
@OR13 You're right, the `getContextFromIntermediate` function in the [linked test](https://github.com/transmute-industries/verifiable-data/blob/main/packages/jsonld-schema/src/__tests__/duplicate-terms.test.ts) does the trick!

It looks as though the issue is that while the code defining same-IRI class properties as arrays (the `getIntermediateFromDirectory` function from[ jsonld-schema](https://github.com/transmute-industries/verifiable-data/tree/main/packages/jsonld-schema)) is imported into getIntermediateFromDirectory.js in traceability-vocab, the code responsible for handling these arrays (`getContextFromIntermediate`) is not similarly imported into traceability-vocab's getContextFromIntermediate.js. So, traceability-vocab only has part of your solution.

I can think of two ways to resolve this:

1. Update traceability-vocab's getContextFromIntermediate.js to import the `getContextFromIntermediate` function from jsonld-schema. A small obstacle to this is that whereas the `getIntermediateFromDirectory` function in jsonld-schema is directly exported (under the new name `schemasToIntermediate`), `getContextFromIntermediate` is not similarly exported. Instead, a function `schemasToContext` is imported which calls both `getIntermediateFromDIrectory` AND `getContextFromIntermediate`. So schemasToContext.ts in jsonld-schema would need to be updated as well as getContextFromIntermediate.js in traceability-vocab, in order to export getContextFromIntermediate directly.
2. Integrate both getIntermediateFromDirectory and getContextFromIntermediate from jsonld-schema into traceability-vocab. This would mostly just be copy/paste, and would have the added benefit of simplifying dependencies for traceability-vocab.

I found it a bit confusing navigating between traceability-vocab and jsonld-schema to get the full picture of what's happening from schema to context so I'd personally lean toward 2, but I'd be very interested to hear your take (& others') on the pros/cons.

Relevant code in jsonld-schema:
* https://github.com/transmute-industries/verifiable-data/blob/main/packages/jsonld-schema/src/schemasToContext.ts

Relevant code in traceability-vocab:
* https://github.com/w3c-ccg/traceability-vocab/blob/main/packages/traceability-schemas/scripts/getIntermediateFromDirectory.js
* https://github.com/w3c-ccg/traceability-vocab/blob/main/packages/traceability-schemas/scripts/getContextFromIntermediate.js

cc: @mprorock

**Comment by OR13** [comment]
yes, we should remove the duplicated code from trace vocab.

**Comment by rhofvendahl** [comment]
@OR13 Sure. I can create a PR in transmute-industries/verifiable-data to export `getContextFromIntermediate` and another PR in w3c-ccg/traceability-vocab to remove the existing code in getContextFromIntermediate.js and import+use that instead.

Just so I understand, what's the benefit of placing `getIntermediateFromDirectory` and `getContextFromIntermediate` in an external package?

**Comment by OR13** [comment]
external packages that have CI/CD enabled and that can take PRs faster are a better way of building software... my experience with the CCG has led me to avoid trying to develop software here...  and instead, focus on specs... this repo has a lot of "spec tooling"... which can probably move faster if its managed independently.

**Comment by OR13** [comment]
I believe we fixed this issue, and if not, we should address any remaining type issues on a case by case basis.
